presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -32,319 +32,206 @@
 import static com.google.common.base.Verify.verify;
 import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
-import static io.airlift.slice.SizeOf.SIZE_OF_INT;
 import static java.util.Objects.requireNonNull;
 
 public class JtsGeometrySerde
 {
-    /**
-     * Shape type codes from ERSI's specification
-     * https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf
-     */
-    private enum EsriShapeType
-    {
-        POINT(1),
-        POLYLINE(3),
-        POLYGON(5),
-        MULTI_POINT(8);
-
-        final int code;
-
-        EsriShapeType(int code)
-        {
-            this.code = code;
-        }
-
-        static EsriShapeType valueOf(int code)
-        {
-            for (EsriShapeType type : values()) {
-                if (type.code == code) {
-                    return type;
-                }
-            }
-
-            throw new IllegalArgumentException("Unsupported ESRI shape code: " + code);
-        }
-    }
-
-    /**
-     * Shape types defined by JTS.
-     */
-    private static final String JTS_POINT = "Point";
-    private static final String JTS_POLYGON = "Polygon";
-    private static final String JTS_LINESTRING = "LineString";
-    private static final String JTS_MULTI_POINT = "MultiPoint";
-    private static final String JTS_MULTI_POLYGON = "MultiPolygon";
-    private static final String JTS_MULTI_LINESTRING = "MultiLineString";
-    private static final String JTS_GEOMETRY_COLLECTION = "GeometryCollection";
-
+    // TODO: Are we sure this is thread safe?
     private static final GeometryFactory GEOMETRY_FACTORY = new GeometryFactory();
 
     private JtsGeometrySerde() {}
 
-    /**
-     * Deserializes ESRI shape as described in
-     * https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf
-     * into JTS Geometry object.
-     */
     public static Geometry deserialize(Slice shape)
     {
-        if (shape == null) {
-            return null;
-        }
-
+        requireNonNull(shape, "shape is null");
         BasicSliceInput input = shape.getInput();
-        int geometryType = input.readByte();
-        if (geometryType != GeometryType.GEOMETRY_COLLECTION.code()) {
-            return readGeometry(input);
-        }
-
-        if (input.available() == 0) {
-            return GEOMETRY_FACTORY.createGeometryCollection();
-        }
-
-        // GeometryCollection: geometry-type|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
-        List<Geometry> geometries = new ArrayList<>();
-        while (input.available() > 0) {
-            int length = input.readInt();
-            geometries.add(readGeometry(input.readSlice(length).getInput()));
-        }
-
-        return GEOMETRY_FACTORY.createGeometryCollection(geometries.toArray(new Geometry[0]));
+        verify(input.available() > 0);
+        GeometryType type = GeometryType.getForCode(input.readByte());
+        return readGeometry(input, type);
```

Comment:
You can inline `type` here.

Reply:
I'm not a big fan of inlining any calls related to a stream reads. It is harder to debug than, since you can't read from a stream twice. So it is always nice to have a separate variable, to be able to stop with a debugger and check what has been read.